crates/red_knot_python_semantic/src/visibility_constraints.rs

```diff
@@ -182,211 +176,389 @@ const MAX_RECURSION_DEPTH: usize = 24;
 /// That means that when you are constructing a formula, you might need to create distinct atoms
 /// for a particular [`Constraint`], if your formula needs to consider how a particular runtime
 /// property might be different at different points in the execution of the program.
-#[derive(Clone, Debug, PartialEq, Eq)]
-pub(crate) struct VisibilityConstraint<'db>(VisibilityConstraintInner<'db>);
-
-#[derive(Clone, Debug, PartialEq, Eq)]
-pub(crate) enum VisibilityConstraintInner<'db> {
-    AlwaysTrue,
-    AlwaysFalse,
-    Ambiguous,
-    VisibleIf(Constraint<'db>, u8),
-    VisibleIfNot(ScopedVisibilityConstraintId),
-    KleeneAnd(ScopedVisibilityConstraintId, ScopedVisibilityConstraintId),
-    KleeneOr(ScopedVisibilityConstraintId, ScopedVisibilityConstraintId),
+///
+/// Visibility constraints are normalized, so equivalent constraints are guaranteed to have equal
+/// IDs.
+#[derive(Clone, Copy, Eq, Hash, PartialEq)]
+pub(crate) struct ScopedVisibilityConstraintId(u32);
+
+impl std::fmt::Debug for ScopedVisibilityConstraintId {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        let mut f = f.debug_tuple("ScopedVisibilityConstraintId");
+        match *self {
+            ALWAYS_TRUE => f.field(&format_args!("AlwaysTrue")).finish(),
+            AMBIGUOUS => f.field(&format_args!("Ambiguous")).finish(),
+            ALWAYS_FALSE => f.field(&format_args!("AlwaysFalse")).finish(),
+            _ => f.field(&self.0).finish(),
+        }
+    }
+}
+
+// Internal details:
+//
+// There are 3 terminals, with hard-coded constraint IDs: true, ambiguous, and false.
+//
+// _Atoms_ are the underlying Constraints, which are the variables that are evaluated by the
+// ternary function.
+//
+// _Interior nodes_ provide the TDD structure for the formula. Interior nodes are stored in an
+// arena Vec, with the constraint ID providing an index into the arena.
+
+#[derive(Clone, Copy, Debug, Eq, Hash, PartialEq)]
+struct InteriorNode {
+    atom: Atom,
+    if_true: ScopedVisibilityConstraintId,
+    if_ambiguous: ScopedVisibilityConstraintId,
+    if_false: ScopedVisibilityConstraintId,
+}
+
+/// A "variable" that is evaluated as part of a TDD ternary function. For visibility constraints,
+/// this is (one of the copies of) a `Constraint` that represents some runtime property of the
+/// Python code that we are evaluating. We intern these constraints in an arena
+/// ([`VisibilityConstraints::constraints`]). An atom consists of an index into this arena, and a
+/// copy number.
```

Comment:
It's unclear to me what a copy number is. Can you expand the comment to cover this in more detail. It may also be worthwhile to explain the internal representation. How many constraints can be expressed etc.

Reply:
Done